spotbugs/src/main/java/edu/umd/cs/findbugs/detect/DateFormatStringChecker.java

```diff
@@ -0,0 +1,129 @@
+import org.apache.bcel.Const;
+
+import edu.umd.cs.findbugs.BugInstance;
+import edu.umd.cs.findbugs.BugReporter;
+import edu.umd.cs.findbugs.StringAnnotation;
+import edu.umd.cs.findbugs.annotations.Nullable;
+import edu.umd.cs.findbugs.annotations.NonNull;
+import edu.umd.cs.findbugs.bcel.OpcodeStackDetector;
+
+import java.util.Arrays;
+import java.util.List;
+
+public class DateFormatStringChecker extends OpcodeStackDetector {
+    private final BugReporter bugReporter;
+
+    /**
+     * Stores required and forbidden flags. Then confirms the absence of a forbidden
+     * flag and presence of all required flags via verify method
+     */
+    private static final class DateFormatRule {
+
+        /**
+         * forbiddenFlag must be missing to match given DateFormatRule. If this field
+         * is {@code null}, rule does not have a forbiddenFlag and the associated check
+         * will be bypassed.
+         */
+        @Nullable
+        String forbiddenFlag;
+
+        /**
+         * List of flags required in a pattern to match given DateFormatRule.
+         */
+        @NonNull
+        List<String> requiredFlags;
+
+        DateFormatRule(String forbiddenFlag, String... requiredFlags) {
+            this.forbiddenFlag = forbiddenFlag;
+            this.requiredFlags = Arrays.asList(requiredFlags);
+        }
+
+        /**
+         * Checks if string does not contain a forbiddenFlag and has all requiredFlags
+         *
+         * @param dateFormatString - string to be checked
+         * @return {@code true} if code equals to one of the elements
+         */
+        boolean verify(String dateFormatString) {
+            if (this.forbiddenFlag != null && dateFormatString.contains(this.forbiddenFlag)) {
+                return false;
+            }
+            for (String flag : requiredFlags) {
+                if (!dateFormatString.contains(flag)) {
+                    return false;
+                }
+            }
+            return true;
+        }
+    }
+
+    public DateFormatStringChecker(BugReporter bugReporter) {
+        this.bugReporter = bugReporter;
+    }
+
+    @Override
+    public void sawOpcode(int seen) {
+        if (!(seen == Const.INVOKESPECIAL || seen == Const.INVOKEVIRTUAL || seen == Const.INVOKESTATIC
+                || seen == Const.INVOKEINTERFACE) || stack.getStackDepth() == 0) {
+            return;
+        }
+
+        if (Arrays.asList(Const.CONSTRUCTOR_NAME, "applyPattern", "applyLocalizedPattern")
+                .contains(getNameConstantOperand())
+                && getSigConstantOperand().contains("Ljava/lang/String;")
+                && "java/text/SimpleDateFormat".equals(getClassConstantOperand())) {
+
+            String dateFormatString = (String) stack.getStackItem(1).getConstant();
+            if (new SignatureParser(getSigConstantOperand()).getNumParameters() == 1) {
+                dateFormatString = (String) stack.getStackItem(0).getConstant();
+            }
```

Comment:
fixed: it uses position of latest parameter in the stack (as the order is reversed)

Reply:
It's still not easy to get why that is the relevant parameter. That is only a lucky coincidence, that it's every time that parameter. It can be different for different classes and functions. Can you please check the exact signatures (or at least the parameter parts) for the different functions and set this `dateFormatString` variable according to that?